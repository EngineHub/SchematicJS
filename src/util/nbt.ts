import {
    Byte,
    Float,
    getTagType,
    Int,
    Short,
    TagArray,
    TagMap,
    TagObject,
    TagType
} from 'nbt-ts';

export function normaliseCompoundToObject(tag: TagObject | TagMap): TagObject {
    if (tag instanceof Map) {
        tag = Object.fromEntries(tag);
    }

    return tag;
}

export function normaliseCompoundToMap(tag: TagObject | TagMap): TagMap {
    if (tag.constructor === Object) {
        tag = new Map(Object.entries(tag)) as TagMap;
    }

    return tag as TagMap;
}

export function tagToRecord(tag: TagObject | TagMap): Record<string, unknown> {
    const metadata: Record<string, unknown> = {};

    const tagObj = normaliseCompoundToObject(tag);

    Object.keys(tagObj).forEach(key => {
        const value = tagObj[key];
        switch (getTagType(value)) {
            case TagType.Byte:
            case TagType.Short:
            case TagType.Int:
            case TagType.Float:
                metadata[key] = (value as Byte | Short | Int | Float).value;
                break;
            case TagType.Long:
            case TagType.Double:
            case TagType.String:
            case TagType.ByteArray:
            case TagType.IntArray:
            case TagType.LongArray:
                metadata[key] = value;
                break;
            case TagType.Compound:
                metadata[key] = tagToRecord(value as TagMap | TagObject);
                break;
            case TagType.List:
                metadata[key] = tagToArray(value as TagArray);
                break;
            default:
                throw new Error(`Unknown tag type: ${getTagType(value)}`);
        }
    });

    return metadata;
}

export function tagToArray(tag: TagArray): unknown[] {
    const array: unknown[] = [];

    tag.forEach((value, index) => {
        switch (getTagType(value)) {
            case TagType.Byte:
            case TagType.Short:
            case TagType.Int:
            case TagType.Float:
                array[index] = (value as Byte | Short | Int | Float).value;
                break;
            case TagType.Long:
            case TagType.Double:
            case TagType.String:
            case TagType.ByteArray:
            case TagType.IntArray:
            case TagType.LongArray:
                array[index] = value;
                break;
            case TagType.Compound:
                array[index] = tagToRecord(value as TagMap);
                break;
            case TagType.List:
                array[index] = tagToArray(value as TagArray);
                break;
            default:
                throw new Error(`Unknown tag type: ${getTagType(value)}`);
        }
    });

    return array;
}
