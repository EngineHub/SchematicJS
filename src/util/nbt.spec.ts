import { Byte, Short, TagMap } from 'nbt-ts';
import { tagToRecord } from './nbt';

describe('tagToRecord', () => {
    it('should convert primitives', () => {
        const tag = new Map<string, unknown>([
            ['byte', new Byte(1)],
            ['short', new Short(2)],
            ['string', 'string']
        ]) as TagMap;

        expect(tagToRecord(tag)).toMatchSnapshot();
    });

    it('should convert arrays', () => {
        const tag = new Map<string, unknown>([
            ['byte', [new Byte(1), new Byte(2)]],
            ['short', [1, 2, 3]],
            ['string', ['string', 'string2']]
        ]) as TagMap;

        expect(tagToRecord(tag)).toMatchSnapshot();
    });

    it('should convert objects', () => {
        const tag = new Map<string, unknown>([
            ['byte', { a: new Byte(1), b: new Byte(2) }],
            ['string', { a: 'string', b: 'string2' }]
        ]) as TagMap;

        expect(tagToRecord(tag)).toMatchSnapshot();
    });
});
