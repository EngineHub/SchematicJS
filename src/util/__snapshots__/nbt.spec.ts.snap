// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tagToRecord should convert arrays 1`] = `
Object {
  "byte": Array [
    1,
    2,
  ],
  "short": Array [
    1,
    2,
    3,
  ],
  "string": Array [
    "string",
    "string2",
  ],
}
`;

exports[`tagToRecord should convert objects 1`] = `
Object {
  "byte": Object {
    "a": 1,
    "b": 2,
  },
  "string": Object {
    "a": "string",
    "b": "string2",
  },
}
`;

exports[`tagToRecord should convert primitives 1`] = `
Object {
  "byte": 1,
  "short": 2,
  "string": "string",
}
`;
